### DISCUSSION TITLE
Doubts with grouping elements

### DISCUSSION THREAD

**Discussion by VADDIJASWANT** [clientIO/joint#2410] [Q&A] [answered]
### Introduction

Hi everyone,

I've encountered an issue while working on my project, specifically related to grouping and ungrouping elements. I found inspiration in this CodePen: [JointJS CodePen](https://codepen.io/jointjs/pen/yLvaVrJ?editors=0010).

After successfully implementing grouping in my project, I ran into difficulties when attempting to ungroup elements. The problem lies in the fact that I can't seem to see the handles for selection after ungrouping.

Here's the crucial code snippet where I handle the click event on a paper element:
`
function onPaperElementClick(elementView, evt) {

    pub.set_settings_div_position(elementView);

    const { model } = elementView;

    paper.removeTools();

    const [group] = model.getAncestors().reverse();

    if (group) {
        selection.collection.add([group]);
    } else if (evt.ctrlKey || evt.metaKey) {
        selection.collection.add(elementView.model);
    }

    const halo = new joint.ui.Halo({
        cellView: elementView,
        type: "surrounding",
        boxContent: false
    });

    halo.render();
}
`

In the line const [group] = model.getAncestors().reverse();, I expected to get an array of elements embedded in the group, similar to the CodePen reference. However, I'm only getting one element, not an array. I tried using const group = model.getEmbeddedCells(), but it resulted in a TypeError.

I'm seeking guidance on resolving this issue. How can I correctly retrieve an array of embedded elements when ungrouping?

Any help or insights would be greatly appreciated. You can refer to the attached image for the specific TypeError I'm encountering.

Thank you

![image](https://github.com/clientIO/joint/assets/116190897/d6de0308-cfb2-4423-8cfb-f134632235a6)
https://github.com/clientIO/joint/assets/116190897/5c8ca3ac-f79b-4375-a7e2-495f8f9827f5

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
```
el.getAncestors();
```
Returns an array of `el`'s parent, the parent of this parent, etc.

```
el.getEmbeddedCells({ deep: true })
```
Returns all embedded elements including any level of nesting.

---
> I tried using const group = model.getEmbeddedCells(), but it resulted in a TypeError.

What type error?

---

The error you have shared is related to another issue: https://github.com/clientIO/joint/discussions/2314

**Reply by VADDIJASWANT** [reply]
Hi @kumilingus ,In my project, I'm not using npm, and my Backbone script file is specifically version 1.4.1. Just to clarify, I obtained this file from Rappid v3.7.2, and I haven't made any modifications to any of the dependency files.

Considering this, do you think the issue might be arising from a different source, or should I still look into adjusting the Backbone version?

**Reply by kumilingus** [reply]
For starters, I would like to find out if `model.getEmbeddedCells()` and `"element.getAncestors is not function"` are related or not. Shouldn't they be 2 different tickets?

**Reply by VADDIJASWANT** [reply]
Let me clarify the issue I'm encountering. The main challenge is that I'm unable to get handles for the selection UI when dealing with grouped elements.

Normally, selection handles appear on the UI only if there are multiple elements in selection.collection. Referencing the CodePen I shared earlier, I noticed that clicking on an element provides an array of embedded elements using element.getAncestors(). I modified my element.click function in my project accordingly to gather all embedded elements and pass them to selection.collection.add().

However, when using element.getAncestors(), I'm only getting one element in return instead of all the elements in its group. I then attempted to use element.getEmbeddedCells(), but encountered type errors (attached).

So, my query revolves around both of these methods. Still, my primary concern remains adding handles to the selection UI for grouped elements. I've implemented a custom function in the handles to ungroup elements, but the UI handles for grouped elements are not showing up.

Any guidance on this would be immensely helpful.

**Reply by kumilingus** [reply]
Calling `element.getEmbeddedCells()` can not throw `element.getAncestors is not function` error.

The `element.getAncestors is not function` comes from this line:
```js
selection.collection.add([group]);
```

It should be:
```js
selection.collection.add(group);
```
`.add` method expects a single model.

---

Now you should be able get all the embedded children using `element.getEmbeddedCells()` and continue.

**Reply by VADDIJASWANT** [reply]
Hi @kumilingus,

Thanks for your previous reply – it helped resolve the type error issue I was encountering. However, the primary issue I'm facing still persists. I realized I might not have clearly explained my doubt. To make things clearer, I've created a [CodePen](https://codepen.io/VADDIJASWANT/pen/RwvYmNP) replicating the problem.

In this CodePen, when you group elements, you'll notice that ungrouping is problematic as the selection UI handle doesn't appear. Could you please take a look and help clarify my doubt?

Thanks!

**Reply by kumilingus** [reply, marked as answer]
You are missing one CSS rule (selection wrapper and handles are hidden by default for single element selections).
```css
.selection-wrapper[data-selection-length="1"] {
  display: block !important;
}
```
